### PR TITLE
HPCC-13724 Review use of batch in cassandra plugin

### PR DESCRIPTION
**Pull request by richardkchapman**
If batch mode not specified on a bulk insert, use async futures to ensure
optimal performance.

Can set MAXFUTURES=nn option to control how many futures will be used, default
is 100.

Signed-off-by: Richard Chapman rchapman@hpccsystems.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday @boneill42 Please review

File: plugins/cassandra/cassandraembed.cpp [hunk 1]
```diff
@@ -243,11 +243,91 @@ class CassandraStatement : public CInterface
...
+                return true;
+        }
+        return false;
+    }
+    void execute()
+    {
+        if (limiter)
+            limiter->wait();
+        future = cass_session_execute(session, statement);
+        if (limiter)
```
Comment (ghalliday) [inline plugins/cassandra/cassandraembed.cpp]: is there a small window for a result to come back before the callback is registered?

File: plugins/cassandra/cassandraembed.cpp [hunk 2]
```diff
@@ -243,11 +243,91 @@ class CassandraStatement : public CInterface
...
 private:
     CassandraStatement(const CassandraStatement &);
     CassStatement *statement;
 };
 
+class CassandraRetryingFuture : public CInterface
+{
+public:
+    CassandraRetryingFuture(CassSession *_session, CassStatement *_statement, Semaphore *_limiter = NULL, unsigned _retries = 10)
+    : session(_session), statement(_statement), retries(_retries), limiter(_limiter)
```
Comment (ghalliday) [inline plugins/cassandra/cassandraembed.cpp]: minor: *slightly* cleaner if future was initialised to NULL.

File: plugins/cassandra/cassandraembed.cpp (hunk 1 above)
Comment (richardkchapman) [inline plugins/cassandra/cassandraembed.cpp]: Maybe, but I don't know what I can do to mitigate it. Any suggestions?

Depends I suppose on the implementation of the cassandra library - will it call back to signal the state change as soon as the callback is registered.

Comment (richardkchapman) [inline plugins/cassandra/cassandraembed.cpp]: Looks like it does:

bool Future::set_callback(Future::Callback callback, void* data) {
  ScopedMutex lock(&mutex_);
  if (callback_) {
    return false; // Callback is already set
  }
  callback_ = callback;
  data_ = data;
  if (is_set_) {
    // Run the callback if the future is already set
    lock.unlock();
    callback(CassFuture::to(this), data);
  }
  return true;
}

**Pull request merged**